Sources/BSON/Integers+BSONValue.swift

```diff
@@ -1,6 +1,52 @@
 import NIO
 
 extension Int32: BSONValue {
+    /*
+     * Initializes an `Int32` from ExtendedJSON
+     *
+     * Parameters:
+     *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for an `Int32`.
+     *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+     *              This is used for error messages.
+     *
+     * Returns:
+     *   - `nil` if the provided value is not an `Int32`.
+     *
+     * Throws:
+     *   - `DecodingError` if `json` is a partial match or is malformed.
+     */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]) throws {
+        switch json {
+        case let .number(n):
+            // relaxed extended JSON
+            guard let int = Int32(exactly: n) else {
+                return nil
+            }
+            self = int
+        case let .object(obj):
+            // canonical extended JSON
+            guard let value = obj["$numberInt"] else {
+                return nil
+            }
+            guard obj.count == 1 else {
+                throw DecodingError._extendedJSONError(
+                    keyPath: keyPath,
+                    debugDescription: "Expected only \"$numberInt\" key, found extra keys in object: \(obj)"
```

Comment:
this will include the entire object, but we only really care about the extra keys. To keep the error message brief, we can just do `obj.keys` here.

Reply:
okay, I changed the wording a bit to make it a bit clearer that we are showing the keys